quiz_app/serializers.py

```diff
@@ -0,0 +1,219 @@
+        if not quiz:
+            raise serializers.ValidationError({'error': 'Invalid quiz ID'})
+
+        question = Question.objects.create(quiz_id=quiz_id, **validated_data)
+
+        for answer_data in answers_data:
+            Answer.objects.create(question=question, **answer_data)
+
+        return question
+
+    def update(self, instance, validated_data):
+        instance.question_text = validated_data.get('question_text', instance.question_text)
+        instance.question_type = validated_data.get('question_type', instance.question_type)
+        instance.point = validated_data.get('point', instance.point)
+        instance.save()
+
+        return instance
+
+
+class QuizSerializer(serializers.ModelSerializer):
+    questions = QuestionSerializer(many=True)
+
+    class Meta:
+        model = Quiz
+        fields = ('id', 'title', 'description', 'time_limit', 'tags', 'categories', 'owner', 'questions')
+        read_only_fields = ['owner', ]
+
+    def create(self, validated_data):
+        questions_data = validated_data.pop('questions')
+        tags = validated_data.pop('tags')
+        categories = validated_data.pop('categories')
+
+        quiz = Quiz.objects.create(**validated_data)
+        quiz.tags.set(tags)
+        quiz.categories.set(categories)
+
+        for question_data in questions_data:
+            answers_data = question_data.pop('answers')
+            question = Question.objects.create(quiz=quiz, **question_data)
+
+            for answer_data in answers_data:
+                Answer.objects.create(question=question, **answer_data)
+
+        return quiz
+
+    def update(self, instance, validated_data):
+        instance.title = validated_data.get('title', instance.title)
+        instance.description = validated_data.get('description', instance.description)
+        instance.time_limit = validated_data.get('time_limit', instance.time_limit)
+
+        tags_data = validated_data.get('tags', instance.tags)
+        instance.tags.set(tags_data)
+
+        categories_data = validated_data.get('categories', instance.categories)
+        instance.categories.set(categories_data)
+
+        instance.save()
+
+        return instance
+
+
+class SubmitAnswerSerializer(serializers.Serializer):
+    question_id = serializers.IntegerField()
+    selected_answer = serializers.IntegerField()
+
+    def validate(self, data):
+        question_id = data.get('question_id')
+        selected_answer = data.get('selected_answer')
+
+        try:
+            question = Question.objects.get(id=question_id)
+            Answer.objects.get(id=selected_answer, question=question)
+        except Question.DoesNotExist:
+            raise serializers.ValidationError("Invalid question ID")
+        except Answer.DoesNotExist:
+            raise serializers.ValidationError("Invalid selected answer / answer is not belong to the given question")
+
+        return data
+
+
+class SubmitQuizSerializer(serializers.Serializer):
```

Comment:
score calculation should be a utility function that can be called on model signals. if that function takes time to finish then use call it as a background task.